### ISSUE TITLE
Help testing code changes in ErlangCompilerError

### ISSUE THREAD

**Issue by martinpritchard** [ignatov/intellij-erlang#796]
Hi, I'm attempting to analyse and hopefully fix some issues that I'm having with erlang compilation errors/warnings, but I'm having difficulty getting any changes that I make take effect when using the 'Erlang' run configuration target.  

I'm making changes to class org.intellij.erlang.jps.builder.ErlangCompilerError, I've added log statements, coded throwing of Exceptions, etc, but no changes that I make seem to have any effect.

I've come to the conclusion that my code changes in this area are not being loaded into the running instance of the IDE.  However, if I change the description in plugin.xml I can see this change in Settings/Plugins, so I know that some changes are at least effective.

Is there something that I'm missing?  Some build target that needs running?  I'd also like to add -verbose:class to see where classes are being loaded from, but I can't seem to find the right place to add this.  I'm new to gradle so I may be missing something.

Thanks in advance.

**Comment by ignatov** [comment]
We're using a standalone process for running JPS builder. Please specify `compiler.process.debug.port` = 5050 in Registry or something like that and start a remote debug session for that process.

1. Go to action via Cmd + Shift + A or Ctrl + Shift +A on Linux/Windows.
2. Find registry action
3. Find compiler.process.debug.port and set some value, e.g. 5050
4. Create Remote Debug Configuration and use 5050
5. Debug with it

**Comment by martinpritchard** [comment]
OK thanks.  Here's what I've done to get a debugging session:
1. Run the Erlang Run configuration to get a new intellij session with some erlang code
2. In this new node Ctrl+Shift+A and set compiler.process.debug.port
3. In this new node Ctrl+Shift+A and set DebugBuildProess to ON
4. Change the erlang code so that it doesn't compile.  Red warning lines appear in the IDE.
5. Goto Build/Build Project, the build suspends waiting for remote debugging session to Attach
6. Go back to the first intellij session and run a remote debug session with the correct port
7. Step through the code, etc

Now I have a debugging session that runs, but I have a few questions:
1. When I run Build/Build Project, compilation always succeeds, even when there are compilation errors that are highlighted in the IDE, I have also copied and pasted the command that is reported to be run and it shows up syntax errors.
2. When I step through the code I can see the ErlangBuilder constructor being called, but the build() method is not executed.  This is probably related to the above, maybe my project setup is wrong?
3. How do I debug and change the compilation that is done in the IDE whilst typing, i.e. without hitting Build/Build Project?  I'm interested in making improvements in this area.

Thanks

**Comment by ignatov** [comment]
@chashnikov could you please advice?

**Comment by chashnikov** [comment]
@martinpritchard
2. ErlangBuilder is a TargetBuilder which builds instances of ErlangTarget (see javadoc of BuildTarget for details). So if its 'build' method isn't called maybe there're no ErlangTargets in the project, e.g. because there're no Erlang modules in the project (see how [ErlangTargetType.computeAllTargets](https://github.com/ignatov/intellij-erlang/blob/master/jps-plugin/src/org/intellij/erlang/jps/builder/ErlangTargetType.java#L41) is implemented). Or maybe it isn't included into CompileScope. You can put a breakpoint into org.jetbrains.jps.incremental.IncProjectBuilder#buildChunkIfAffected method and debug it to check why your target isn't processed.
3. The IDE doesn't perform compilation to highlight errors in the editor, this is implemented via [Inspections](http://www.jetbrains.org/intellij/sdk/docs/tutorials/code_inspections.html).

**Comment by martinpritchard** [comment]
thanks @chashnikov this gives me a bit more to go on, I'll take a look.

**Comment by martinpritchard** [comment]
Hi @chashnikov, I've taken a look at the Inspections used, but would you be able to help me understand why erlc isn't involved in highlighting compilation errors as you type?  There are bugs in this area that I'd like to help resolve - I sometimes push from within the IDE and CI picks up compilation errors.  My first thought is to call erlc and use that as an Annotator or External Annotator - is this feasible or would this cause issues?

**Comment by chashnikov** [comment]
@martinpritchard Yes, it's possible to invoke a compiler via ExternalAnnotator. However it may lead to problems when it compile a file which depends on another file.
